# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (20, 30)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 20 to 30:
china is usually pleasant during november , and it is never quiet in october .
the united states is never nice during february , but it is sometimes pleasant in april .
india is never busy during autumn , and it is mild in spring .
paris is mild during summer , but it is usually busy in april .
france is never cold during september , and it is snowy in october .
california is never cold during may , and it is sometimes chilly in march .
he dislikes lemons , grapes , and mangoes.
their favorite fruit is the mango , but our favorite is the pear .
france is sometimes quiet during may , and it is never chilly in august .
paris is never pleasant during september , and it is beautiful in autumn .

French sentences 20 to 30:
chine est généralement agréable en novembre , et il est jamais tranquille en octobre .
les états-unis e

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of each sentence from `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def transform_vocab(text, vocab):
    sentences = []
    for sentence in text.split('\n'):
        ids = []
        for word in sentence.split():
            ids.append( vocab[word] )
        
        sentences.append(
            ids
        )
    return sentences

def add_eos(sentences, token):
    for s in sentences:
        s.append(token)
    return sentences

def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    source_sentences = transform_vocab(source_text, source_vocab_to_int)
    target_sentences = transform_vocab(target_text, target_vocab_to_int)
    target_sentenEOS = add_eos(target_sentences, target_vocab_to_int['<EOS>'])
    
    return source_sentences, target_sentenEOS



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf

# Check TensorFlow Version
assert LooseVersion(tf.__version__) in [LooseVersion('1.0.0'), LooseVersion('1.0.1')], 'This project requires TensorFlow version 1.0  You are using {}'.format(tf.__version__)
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.0.1
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoding_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.

Return the placeholders in the following the tuple (Input, Targets, Learing Rate, Keep Probability)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, and learning rate.
    :return: Tuple (input, targets, learning rate, keep probability)
    """
    _input = tf.placeholder(shape=[None, None],dtype=tf.int32, name="input")
    targets = tf.placeholder(shape=[None, None], dtype=tf.int32)
    learning_rate = tf.placeholder(shape=[], dtype=tf.float32)
    keep_probability = tf.placeholder(shape=[], dtype=tf.float32, name="keep_prob")
    
    return _input, targets, learning_rate, keep_probability



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoding Input
Implement `process_decoding_input` using TensorFlow to remove the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoding_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for dencoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    dec_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    return dec_input


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_decoding_input(process_decoding_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer using [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn).

In [9]:
def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :return: RNN state
    """
        
    # source: Google devs - Youtube: Sequence Models and the RNN API (TFsummit) 
    # devices = ["/gpu:1","/gpu:0"]   # layer per device
    # cell = MultiRnnCell([Devicewrapper(ResidualWrapper(LSTMCell(rnnsize)), device) for device in devices])
    
    basic_cell = tf.contrib.rnn.BasicLSTMCell(rnn_size)    
    multi_cell = tf.contrib.rnn.MultiRNNCell([basic_cell] * num_layers)
    dropo_cell = tf.contrib.rnn.DropoutWrapper(multi_cell, input_keep_prob= keep_prob) 
        
    outputs, state = tf.nn.dynamic_rnn(dropo_cell, rnn_inputs, dtype=tf.float32)
    return state


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create training logits using [`tf.contrib.seq2seq.simple_decoder_fn_train()`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/simple_decoder_fn_train) and [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder).  Apply the `output_fn` to the [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder) outputs.

In [10]:
def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, sequence_length, decoding_scope,
                         output_fn, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param sequence_length: Sequence Length
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_fn: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: Train Logits
    """
    train_decoder_fn = tf.contrib.seq2seq.simple_decoder_fn_train(encoder_state)
    
    dec_cell = tf.contrib.rnn.DropoutWrapper(dec_cell, input_keep_prob=keep_prob) 

    train_pred, _, _ = tf.contrib.seq2seq.dynamic_rnn_decoder(
        dec_cell, train_decoder_fn, dec_embed_input, sequence_length, scope=decoding_scope)   
        
    return output_fn(train_pred)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference logits using [`tf.contrib.seq2seq.simple_decoder_fn_inference()`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/simple_decoder_fn_inference) and [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder). 

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id,
                         maximum_length, vocab_size, decoding_scope, output_fn, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param maximum_length: The maximum allowed time steps to decode
    :param vocab_size: Size of vocabulary
    :param decoding_scope: TensorFlow Variable Scope for decoding
    :param output_fn: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: Inference Logits
    """
    d_o_cell = tf.contrib.rnn.DropoutWrapper(dec_cell, input_keep_prob=keep_prob) 
    
    inter_decoder_fn = tf.contrib.seq2seq.simple_decoder_fn_inference(
        output_fn, encoder_state, dec_embeddings, start_of_sequence_id, 
        end_of_sequence_id, maximum_length, vocab_size)
    
    train_pred, _, _ = tf.contrib.seq2seq.dynamic_rnn_decoder(
        d_o_cell, inter_decoder_fn, scope = decoding_scope)
    
    return train_pred


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

- Create RNN cell for decoding using `rnn_size` and `num_layers`.
- Create the output fuction using [`lambda`](https://docs.python.org/3/tutorial/controlflow.html#lambda-expressions) to transform it's input, logits, to class logits.
- Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, sequence_length, decoding_scope, output_fn, keep_prob)` function to get the training logits.
- Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, maximum_length, vocab_size, decoding_scope, output_fn, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
#run_on_server = False      # init - no_GPU 1 or 2  var set true if [DeviceWrapper]

def decoding_layer(dec_embed_input, dec_embeddings, encoder_state, vocab_size, sequence_length, rnn_size,
                   num_layers, target_vocab_to_int, keep_prob):
    """
    Create decoding layer
    :param dec_embed_input: Decoder embedded input
    :param dec_embeddings: Decoder embeddings
    :param encoder_state: The encoded state
    :param vocab_size: Size of vocabulary
    :param sequence_length: Sequence Length
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param keep_prob: Dropout keep probability
    :return: Tuple of (Training Logits, Inference Logits)
    """
    start_of_sequence_id = target_vocab_to_int['<GO>']
    end_of_sequence_id = target_vocab_to_int['<EOS>']
    basic_cell = tf.contrib.rnn.BasicLSTMCell(rnn_size)
    multi_cell = tf.contrib.rnn.MultiRNNCell([basic_cell] * num_layers)
    dropo_cell = tf.contrib.rnn.DropoutWrapper(multi_cell, input_keep_prob= keep_prob) 
        
    output_fn = lambda logits: tf.contrib.layers.fully_connected(inputs=logits, num_outputs=vocab_size, \
                                                                activation_fn=None, scope=decoding_scope)
        
    with tf.variable_scope("decoding") as decoding_scope:
        t_logits = decoding_layer_train(encoder_state, dropo_cell, dec_embed_input, sequence_length, \
                                        decoding_scope, output_fn, keep_prob)
    
    with tf.variable_scope("decoding", reuse=True) as decoding_scope:
        i_logits = decoding_layer_infer(encoder_state, dropo_cell, dec_embeddings, start_of_sequence_id, \
                                        end_of_sequence_id, sequence_length, vocab_size, decoding_scope, \
                                        output_fn, keep_prob)
    
    return t_logits, i_logits
    

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Apply embedding to the input data for the encoder.
- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob)`.
- Process target data using your `process_decoding_input(target_data, target_vocab_to_int, batch_size)` function.
- Apply embedding to the target data for the decoder.
- Decode the encoded input using your `decoding_layer(dec_embed_input, dec_embeddings, encoder_state, vocab_size, sequence_length, rnn_size, num_layers, target_vocab_to_int, keep_prob)`.

In [13]:
def embed(input_data, vocab_size, embd_siz):
    
    init_embd = tf.random_uniform((vocab_size, embd_siz), -1, 1)
    embedding = tf.Variable(init_embd)
    embed_lookup = tf.nn.embedding_lookup(embedding, input_data)
    
    return embedding, embed_lookup


def seq2seq_model(input_data, target_data, keep_prob, batch_size, sequence_length, source_vocab_size, 
                  target_vocab_size, enc_embedding_size, dec_embedding_size, rnn_size, num_layers, 
                  target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param sequence_length: Sequence Length
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training Logits, Inference Logits)
    """
    
    input_data, input_embed_lookup = embed(input_data, source_vocab_size, enc_embedding_size)
    encoder_state = encoding_layer(input_embed_lookup, rnn_size, num_layers, keep_prob)
    
    target_data = process_decoding_input(target_data, target_vocab_to_int, batch_size)
    target_data, target_embed_lookup = embed(target_data, target_vocab_size, dec_embedding_size)
    t_logits, i_logits = decoding_layer(target_embed_lookup, target_data, encoder_state, target_vocab_size, \
                         sequence_length, rnn_size, num_layers, target_vocab_to_int, keep_prob)

    return  t_logits, i_logits

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability

In [14]:
# Number of Epochs
epochs        = 1               # set to 1 ..for initial run
# Batch Size
batch_size    = 512             # 1024 results in: "ressource exhausted" error      
# RNN Size
rnn_size      = 512             # 'high mean, the model will learn more complex structure' 
# Number of Layers      
num_layers    = 4
# Embedding Size                # "Can represent the number of unique words we can deal with." 
encoding_embedding_size = 256
decoding_embedding_size = 256
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.6
# write a line every n batch
show_every_n_batches = 135

print('Learning Rate:', learning_rate)


Learning Rate: 0.01


### Build the Graph
Build the graph using the neural network you implemented.

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_source_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob = model_inputs()
    sequence_length = tf.placeholder_with_default(max_source_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)
    
    train_logits, inference_logits = seq2seq_model(
        tf.reverse(input_data, [-1]), targets, keep_prob, batch_size, sequence_length, len(source_vocab_to_int), len(target_vocab_to_int),
        encoding_embedding_size, decoding_embedding_size, rnn_size, num_layers, target_vocab_to_int)

    tf.identity(inference_logits, 'logits')
    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            train_logits,
            targets,
            tf.ones([input_shape[0], sequence_length]))

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)

### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import time

def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1]), (0,0)],
            'constant')

    return np.mean(np.equal(target, np.argmax(logits, 2)))

train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]

valid_source = helper.pad_sentence_batch(source_int_text[:batch_size])
valid_target = helper.pad_sentence_batch(target_int_text[:batch_size])

with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch) in enumerate(
                helper.batch_data(train_source, train_target, batch_size)):
            start_time = time.time()
            
            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 sequence_length: target_batch.shape[1],
                 keep_prob: keep_probability})
            
            batch_train_logits = sess.run(
                inference_logits,
                {input_data: source_batch, keep_prob: 1.0})
            batch_valid_logits = sess.run(
                inference_logits,
                {input_data: valid_source, keep_prob: 1.0})
                
            train_acc = get_accuracy(target_batch, batch_train_logits)
            valid_acc = get_accuracy(np.array(valid_target), batch_valid_logits)
            end_time = time.time()
            if (batch_i % show_every_n_batches == 0):
                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.3f}, Validation Accuracy: {:>6.3f}, Loss: {:>6.3f}'
                  .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    0/269 - Train Accuracy:  0.166, Validation Accuracy:  0.245, Loss:  5.883
Epoch   0 Batch  135/269 - Train Accuracy:  0.359, Validation Accuracy:  0.419, Loss:  2.784
Model Trained and Saved


## Continue training ##

In [17]:
#params for continue
# Number of Epochs
epochs = 25


In [18]:
#continue training
cont_time = time.time()

train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]

valid_source = helper.pad_sentence_batch(source_int_text[:batch_size])
valid_target = helper.pad_sentence_batch(target_int_text[:batch_size])

with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    saver.restore(sess,save_path)
    
    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch) in enumerate(
                helper.batch_data(train_source, train_target, batch_size)):
            
            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 sequence_length: target_batch.shape[1],
                 keep_prob: keep_probability})
            
            batch_train_logits = sess.run(
                inference_logits,
                {input_data: source_batch, keep_prob: 1.0})
            batch_valid_logits = sess.run(
                inference_logits,
                {input_data: valid_source, keep_prob: 1.0})
                
            train_acc = get_accuracy(target_batch, batch_train_logits)
            valid_acc = get_accuracy(np.array(valid_target), batch_valid_logits)
            
            if (batch_i % show_every_n_batches == 0):
                line_time = time.time()
                time_pass = line_time - cont_time
                print('time: {:>4.0f} secs Epoch {:>3} Batch {:>4}/{} - Train Accu: {:>6.4f}, Val. Accu: {:>6.4f}, Loss: {:>6.4f}'
                     .format(time_pass, epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    save_time = time.time()
    time_pass = save_time - cont_time
    print('time: {:>4.0f} secs'.format(time_pass))
    print('continue-Model Trained and Saved')

time:    1 secs Epoch   0 Batch    0/269 - Train Accu: 0.3596, Val. Accu: 0.4132, Loss: 1.8282
time:  108 secs Epoch   0 Batch  135/269 - Train Accu: 0.2941, Val. Accu: 0.3484, Loss: 1.4051
time:  213 secs Epoch   1 Batch    0/269 - Train Accu: 0.3480, Val. Accu: 0.4077, Loss: 1.5985
time:  321 secs Epoch   1 Batch  135/269 - Train Accu: 0.3685, Val. Accu: 0.4426, Loss: 1.0633
time:  426 secs Epoch   2 Batch    0/269 - Train Accu: 0.4716, Val. Accu: 0.5128, Loss: 0.8996
time:  533 secs Epoch   2 Batch  135/269 - Train Accu: 0.5877, Val. Accu: 0.6163, Loss: 0.7086
time:  639 secs Epoch   3 Batch    0/269 - Train Accu: 0.6238, Val. Accu: 0.6325, Loss: 0.6031
time:  746 secs Epoch   3 Batch  135/269 - Train Accu: 0.6394, Val. Accu: 0.6696, Loss: 0.5695
time:  852 secs Epoch   4 Batch    0/269 - Train Accu: 0.6929, Val. Accu: 0.7066, Loss: 0.5140
time:  959 secs Epoch   4 Batch  135/269 - Train Accu: 0.6807, Val. Accu: 0.7246, Loss: 0.4600
time: 1065 secs Epoch   5 Batch    0/269 - Train A

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [20]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.


- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [21]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    
    s_lower = sentence.lower()
    # for all words in senten - get the int - if none found set to id-of-UNK
    ids = [ vocab_to_int.get(word, vocab_to_int['<UNK>'])  for word in s_lower.split() ]
    
    return ids 


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [22]:
translate_sentence = 'he saw a old yellow truck .'

"""
# google trans (GT:)
# 'he saw a old yellow truck .' (GT:) Il a vu un vieux caminion jaune
        --- Input
            Word Ids:      [18, 190, 218, 72, 146, 159, 124]
            English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']
        Prediction
            Word Ids:      [214, 107, 177, 109, 74, 295, 354, 118, 1]
            French Words: ['il', 'a', 'vu', 'un', 'petit', 'camion', 'jaune', '.', '<EOS>']
        ---

# 'it was the age of wisdom , it was the age of foolishness ,'
        (GT:) C'etait l'age de la sagresse, c'etait l'age de la sottise,
        ---['it', 'was', 'the', '<UNK>', '<UNK>', '<UNK>', ',', 'it', 'was', 'the', '<UNK>', '<UNK>', '<UNK>', ','] 
        
# used to avoid '<UNK>'
# 'new jersey is nice when it is hot , paris is usually busy in july'
        (GT:) New Jersey est agreable quand il est chaud, Paris est generalement occupe en juillet
        --- Input
            Word Ids:      [173, 149, 199, 192, 116, 180, 199, 145, 134, 26, 199, 179, 49, 126, 169]
            English Words: ['new', 'jersey', 'is', 'nice', 'when', 'it', 'is', 'hot', ',', 'paris', 'is', 'usually', 'busy', 'in', 'july']
        Prediction
            Word Ids:      [129, 66, 128, 349, 230, 25, 57, 329, 302, 214, 128, 150, 226, 25, 37, 118, 1]
            French Words: ['new', 'jersey', 'est', 'généralement', 'chaud', 'en', 'mars', ',', 'et', 'il', 'est', 'jamais', 'agréable', 'en', 'avril', '.', '<EOS>']
        ---

# 'he dislikes autumn in china , it is chilly , sometimes it is snowy , it is never mild'
        (GT:) Il n'aime pas l'automne en Chine, il est froid, il est parfois neigeux, il n'est jamais doux
        --- Input
            Word Ids:      [18, 98, 193, 126, 165, 134, 180, 199, 80, 134, 96, 180, 199, 45, 134, 180, 199, 148, 61]
            English Words: ['he', 'dislikes', 'autumn', 'in', 'china', ',', 'it', 'is', 'chilly', ',', 'sometimes', 'it', 'is', 'snowy', ',', 'it', 'is', 'never', 'mild']
        Prediction
            Word Ids:      [214, 128, 349, 76, 157, 112, 203, 241, 171, 329, 279, 150, 203, 324, 183, 241, 171, 128]
            French Words: ['il', 'est', 'généralement', 'froid', 'au', 'cours', 'de', "l'", 'automne', ',', 'mais', 'jamais', 'de', 'neige', 'à', "l'", 'automne', 'est']
        ---
"""

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('logits:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence], keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in np.argmax(translate_logits, 1)]))
print('  French Words: {}'.format([target_int_to_vocab[i] for i in np.argmax(translate_logits, 1)]))

Input
  Word Ids:      [216, 197, 15, 87, 113, 135, 169]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [45, 285, 339, 229, 192, 348, 134, 76, 1]
  French Words: ['il', 'a', 'vu', 'un', 'vieux', 'camion', 'prochain', '.', '<EOS>']


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.